ata/src/config.rs

```diff
@@ -0,0 +1,250 @@
+#[repr(C)]
+#[derive(Clone, Deserialize, Debug, Serialize, Reflect)]
+#[serde(default)]
+pub struct Config {
+    pub api_key: String,
+    pub model: String,
+    pub max_tokens: i64,
+    pub temperature: f64,
+    pub suffix: Option<String>,
+    pub top_p: f64,
+    pub n: u64,
+    pub stream: bool,
+    pub logprobs: u8,
+    pub echo: bool,
+    pub stop: Vec<String>,
+    pub presence_penalty: f64,
+    pub frequency_penalty: f64,
+    pub best_of: u64,
+    pub logit_bias: HashMap<String, f64>,
+}
+
+impl Config {
+    pub fn validate(&self) -> Result<(), String> {
+        if self.api_key.is_empty() {
+            return Err(String::from("API key is missing"));
+        }
+
+        if self.model.is_empty() {
+            return Err(String::from("Model ID is missing"));
+        }
+
+        if self.max_tokens < 1 || self.max_tokens > 2048 {
+            return Err(String::from("Max tokens must be between 1 and 2048"));
+        }
+
+        if self.temperature < 0.0 || self.temperature > 1.0 {
+            return Err(String::from("Temperature must be between 0.0 and 1.0"));
+        }
+
+        if let Some(suffix) = &self.suffix {
+            if suffix.is_empty() {
+                return Err(String::from("Suffix cannot be an empty string"));
+            }
+        }
+
+        if self.top_p < 0.0 || self.top_p > 1.0 {
+            return Err(String::from("Top-p must be between 0.0 and 1.0"));
+        }
+
+        if self.n < 1 || self.n > 10 {
+            return Err(String::from("n must be between 1 and 10"));
+        }
+
+        if self.logprobs > 2 {
+            return Err(String::from("logprobs must be 0, 1, or 2"));
+        }
+
+        if self.stop.iter().any(|stop| stop.is_empty()) || self.stop.len() > 4 {
+            return Err(String::from("Stop phrases cannot contain empties"));
+        }
+
+        if self.presence_penalty < 0.0 || self.presence_penalty > 1.0 {
+            return Err(String::from("Presence penalty must be between 0.0 and 1.0"));
+        }
+
+        if self.frequency_penalty < 0.0 || self.frequency_penalty > 1.0 {
+            return Err(String::from(
+                "Frequency penalty must be between 0.0 and 1.0",
+            ));
+        }
+
+        if self.best_of < 1 || self.best_of > 5 {
+            return Err(String::from("best_of must be between 1 and 5"));
+        }
+
+        for (key, value) in &self.logit_bias {
+            if value < &-2.0 || value > &2.0 {
+                return Err(format!(
+                    "logit_bias for {} must be between -2.0 and 2.0",
+                    key
+                ));
+            }
+        }
+
+        Ok(())
+    }
+}
+
+impl Default for Config {
+    fn default() -> Self {
+        Self {
+            model: "text-davinci-003".into(),
+            max_tokens: 16,
+            temperature: 0.5,
+            suffix: None,
+            top_p: 1.0,
+            n: 1,
+            stream: false,
```

Comment:
Stream has to be enabled to enable streaming. That's one of the main features of `ata`.

Reply:
Maybe I don't want always want streaming and want to pipe to it.

Reply:
> Maybe I don't want always want streaming and want to pipe to it.

How about a Bash script wrapper around `curl`? Something like

```sh
#!/usr/bin/env bash

RESPONSE=$(curl --silent https://api.openai.com/v1/completions \
    -H "Content-Type: application/json" \
    -H "Authorization: Bearer <API KEY>" \
    -d "{\"model\": \"text-davinci-003\", \"prompt\": \"$@\", \"temperature\": 0, \"max_tokens\": 7}")

echo $RESPONSE | jq '.choices[0].text'
```
For example, this gives:
```sh
$ bash tmp.sh 'hi there'
"\n\nHello! How are you"
```

Reply:
I know that's possible but wanted an application to do it.